my-sms/db/migrate/20210212025754_move_student_title_str_to_title_model.rb

```diff
@@ -0,0 +1,44 @@
+class MoveStudentTitleStrToTitleModel < ActiveRecord::Migration
+  def up
+    # cache titles from db to avoid multiple queries
+    title_name_to_id = {}
+    Title.all.each do |title|
+      title_name_to_id[title.name] = title.id
+    end
+
+    # update students table
+    change_table :students do |t|
+      t.rename(:title, :title_old)
+      t.references :title
+    end
+
+    # add title model association
+    # without making assumptions about model
+    # add any titles not already in the db
+    Student.reset_column_information
+    Student.find_each do |student|
+      title = Title.where(name: student[:title_old]).first_or_create
+      student.update_attribute(:title_id, title.id)
+      student.save
+    end
+
+    # remove old title (string) column
+    remove_column :students, :title_old
```

Comment:
If you move the above to a rake task, we could leave this column in. What I've done before is left in columns for at least 1 release, and then removed after. Happy for others to chime in and see what they think though 👍

Reply:
@jhuizy @josh-cooper The pattern Jordan is referring to is common to what we do here at ReadyTech.
The advantage is it acts like an extra safety mechanism.  Deleting Data (ie dropping columns) is highly risky and mitigating the risk where possible is the preferred direction.

If something was missed in the migration code above, and the data in title_old was removed straight away when the column was dropped `remove_column :students, :title_old` the process if getting the backup and restoring 300 clients data in the Production environment is simply not an ideal situation to be in.

If the data was still in the title_old column and something went wrong, its a much cleaner/quicker/safer direction.

Reply:
That makes a lot of sense. I've done both of these things in c5609c85cda148522e852927f980dc94c5ab2025